projen/projen
# fix(file): incorrectly detects file changes on Windows when isExecutable is set

## Motivation

The `package.json` file is marked as executable in a `NodeProject` when running on Windows, but the synthesized one is not executable. Then, this mismatch causes the package manager to rerun every time `projen` runs with unnecessary additional time. It detects correctly if the file is executable on Linux and the issue is only for Windows.

The function that checks is
https://github.com/projen/projen/blob/f8a1ac04d4a635f39a511f2af205dc0a75294d74/src/util.ts#L421-L428

The cause is that it's the same as checking for file existence on Windows, as mentioned in https://nodejs.org/api/fs.html#file-access-constants
>  X_OK: ... . This has no effect on Windows (will behave like fs.constants.F_OK).

The way to correctly detect if a file is executable in Windows is by checking the file extension in the environment variable `PATHEXT`, as [defined in PowerShell help](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.3#path-information).

More generally, this means that all files that are checked by projen for the attribute executable will be marked as such on Windows if they exist when it's not the case.

## Changes

1. Add function `assertExecutablePermissions`.
1. Override File.executable with `assertExecutablePermissions` in File's constructor.
1. Override isExecutable with `assertExecutablePermissions` in util.isExecutable.
1. Unrelated small change: removed a console.log in `test/javascript/eslint.test.ts` that I added accidentally in another PR and got merged.

## Test

1. After the changes in the `projen` repository in WSL locally, run `yarn package:js` to create projen@0.0.0.jsii.tgz
1. Copy projen@0.0.0.jsii.tgz from WSL to Windows.
1. In `my-node-project` in Windows run `yarn install projen@0.0.0.jsii.tgz`
1. Set logging to DEBUG

	```ts
	logging: {
	    level: LogLevel.DEBUG,
	  },
	```

1. Run `yarn projen` several times to check the result.
1. The package manager doesn't run because package.json hasn't changed.
1. Also, the terminal output confirms that package.json is detected as non-executable:
```bash
👾 no change in D:\my-projen-project\package.json
```

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
